keps/sig-node/2570-memory-qos/README.md

```diff
@@ -0,0 +1,297 @@
+
+This proposal uses `memory.min` to protect memory requests for pods and containers. Memory limits are throttled by `memory.high` for `Burstable` containers. We also protect node memory resource tied to `--enforce-node-allocatable`.
+
+| type | qos class | memory.min | memory.high |
+| -------- | -------- | -------- | -------- | 
+| container | guaranteed | pod.spec.containers[i].resources.requests[memory] | n |
+| container | burstable | pod.spec.containers[i].resources.requests[memory] | 85%*pod.spec.containers[i].resources.limits[memory] |
+|  pod | guaranteed / burstable | sum(pod.spec.containers[i].resources.requests[memory]) | n |
+| node | n/a | pods, kube-reserved, system-reserved | n |
+
+### User Stories (Optional)
+#### Memory Sensitive Workload
+Some workloads are sensitive to memory allocation and availability, slight delays may cause service outage. In this case, a mechanism is needed to ensure the quality of memory.
+We must provide guarantee in both of the following aspects:
+- Retain memory requests to reduce allocation latency
+- Protect memory requests from being reclaimed
+
+#### Node Availability
+The stability of kubelet node is very important to users. As the key resource of the node, the availability of memory is the key factor for the stability of the node. We should do something to protect node reserved memory.
+
+### Comparison with Memory Manager
+The Memory Manager is a new component of kubelet ecosystem proposed to enable single-NUMA and multi-NUMA guaranteed memory allocation at topology level. Memory QoS proposal mainly uses cgroups v2 to improve the quality of memory requests, thereby improving the memory qos of `Guaranteed` and `Burstable` pods and even entire node. 
+See also https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1769-memory-manager
+
+### memory.low vs memory.min
+In cgroups v2, `memory.low` is designed for best-effort memory protection which is more like "soft guarantee" and won't be reclaimed unless memory can't be reclaimed from any unprotected cgroups. `memory.min` is a bit aggressive. It will always retain specified amount of memory and it can be never reclaimed. When requirement is not satisfied, system OOM killer will be invoked. 
+
+## Design Details
+### Prerequisite
+1. Kernel enables cgroups v2 unified hierarchy 
+2. CRI runtime supports [cgroups v2 Unified Spec](https://github.com/opencontainers/runtime-spec/blob/7c549cb0939af03d5a2a8b271e2ad6871309e228/specs-go/config.go#L376) for container level
+3. Kubelet enables `--enforce-node-allocatable=<pods, kube-reserved, system-reserved>` 
+
+### Feature Gate
+Set `--feature-gates=MemoryQoS=true` to enable the feature.
+
+### Control Flag
+Set `--enforce-memory-qos=<container, pod, node>` to enable for different level.
```

Comment:
Was this flag recently added? Not sure if I understand the semantics of this. Should it be possible, and will it make sense to eg. use `--enforce-memory-qos=container`? Wouldn't that implicitly mean that all pods and kubepods/node will get `0` as memory.min, making container level value unnecessary for memory.min?

Would be nice with a description about what the values of these flags will mean, what the default is, and what is written to memory.min and memory.high if nothing is set. :smile:

Reply:
Yes. This configuration was intended to control that each level can be enabled individually with memory qos. But according to [latest discussion](https://docs.google.com/document/d/1cXkRkM0XG2le-gbblfvG940puxaPJPYHJh2NTno6p7M/edit?usp=sharing), the proposal should focus on pod/container. I think this flag should be removed first.

Reply:
> [...] the proposal should focus on pod/container.

Not really sure what you mean by that.

> I think this flag should be removed first.

So this flag will be removed from the KEP, or?

Reply:
> > [...] the proposal should focus on pod/container.
> 
> Not really sure what you mean by that.
> 
hmm, the previous design idea of this proposal was to protect memory for all level cgroup in Kubernetes, such as pod, container, qos(guaranteed, burstable) and node(kubepods/kube-reserved/system-reserved). This flag determines which level cgroup should be enabled(Of course, there are other decisive conditions). But according to the latest discussion, we think we should focus on the memory protection of pod/container first. So I think this flag is irrelevant at the moment. Sorry for my poor English :(

> > I think this flag should be removed first.
> 
> So this flag will be removed from the KEP, or?
Yes.

Reply:
> hmm, the previous design idea of this proposal was to protect memory for all level cgroup in Kubernetes, such as pod, container, qos(guaranteed, burstable) and node(kubepods/kube-reserved/system-reserved). This flag determines which level cgroup should be enabled(Of course, there are other decisive conditions). But according to the latest discussion, we think we should focus on the memory protection of pod/container first. So I think this flag is irrelevant at the moment. Sorry for my poor English :(

No problem! :smile: (I am not a native english speaker either, so you are not alone :))

I just we are talking about two separate things when it comes to qos. As I understand memory.min;

If we set `memory.min=0` on `kubepods`, setting `memory.min=200MiB` on a pod cgroup under kubpods will make no difference. Since it will be limited by `kubepods`. Same for `kubepods/kubepods-besteffort`; if that has `memory.min=0`, all pod level `memory.min` inside besteffort will make no difference.

I am not talking about QoS in the sense of prioritization between pods/containers/classes, only that we have to set a value so that we don't make all cgroup memory.min under them do nothing. This is essentially since the default for `memory.min=0` while `memory.limit=unlimited`; so they have to be treated differently.

Reply:
As I read the KEP with the latest update, one has to set `--enforce-node-allocatable=pods` in order to use `memory.min`, and if that is the case, that should probably be stated explicitly.

Reply:
I have re-added qos related contents, and explicitly set `/kubepods/burstable/memory.min=<all-pods-memory-requests>`. Thanks for your correction.